### PR TITLE
[FG:InPlacePodVerticalScaling] Graduate to Beta

### PR DESCRIPTION
**Pull request by tallclair**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

Graduate the `InPlacePodVerticalScaling` feature to beta.

For https://github.com/kubernetes/enhancements/issues/1287

#### Special notes for your reviewer:

Code freeze extension request: https://groups.google.com/g/kubernetes-sig-release/c/DvRlV9mBZ-Q/m/AGmOk6avBwAJ

This should not merge until the remaining beta-blockers are resolved. All of these have PRs inflight.

- [x] https://github.com/kubernetes/kubernetes/issues/117767
- [x] https://github.com/kubernetes/kubernetes/issues/114123
- [x] https://github.com/kubernetes/kubernetes/issues/126388
- [x] https://github.com/kubernetes/kubernetes/issues/127132
- [x] https://github.com/kubernetes/kubernetes/issues/127172 (remaining work is not beta-blocking)
- [x] https://github.com/kubernetes/kubernetes/issues/128068
- [x] https://github.com/kubernetes/kubernetes/issues/119838
- [x] https://github.com/kubernetes/kubernetes/issues/128677

NOTE: The following changes were originally in-scope for beta, but have been moved out:

- https://github.com/kubernetes/kubernetes/issues/128069 - This was always meant to be a best-effort call to inform the runtime of a resize, but the resize is still handled by the Kubelet. This should move to the GA scope, but we do not have a direct consumer or use case for it now.
- https://github.com/kubernetes/kubernetes/issues/128071 - This metric is not sufficiently well defined. With the addition of the `/resize` subresource, we automatically get a metric for total number of resize requests. We will revisit this for GA.
- https://github.com/kubernetes/kubernetes/issues/128070 - Unfortunately this isn't feasible in v1.32 due to a validation rollback issue (see https://github.com/kubernetes/kubernetes/pull/128367/files#r1834897969). We will make the validation changes in v1.32 to make this feasible in v1.33, but the actual implementation won't land until v1.33.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

original release note (reverted in https://github.com/kubernetes/kubernetes/pull/128875)
```
Promoted in-place Pod vertical scaling to beta. The `InPlacePodVerticalScaling` feature gate is now enabled by default.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources
```

/sig node
/priority important-soon
/milestone v1.32

### CODE REVIEW

**Comment by tallclair** [comment]
/hold

Until remaining blockers are closed.

**Comment by kannon92** [comment]
Do we have a periodic that covers this feature?

It’s not blocking but it may be worth having this given the amount of potential regressions.

**Comment by wendy-ha18** [comment]
Hi everyone
A gentle reminder that the code freeze has started 02:00 UTC Friday November 8th 2024 . Please make sure this PR has both lgtm and approved labels ASAP, and file an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md).

**Comment by sftim** [comment]
Changelog suggestion
```diff
-InPlacePodVerticalScaling is now Beta, and enabled by default.
+Promoted in-place Pod vertical scaling to beta. The `InPlacePodVerticalScaling` feature gate is now enabled by default.
```

**Comment by sftim** [comment]
I recommend we propose this to beta but still **off by default**. What do folks think?

My thinking: we should very, very rarely change GA APIs after code freeze, and the case for doing so needs to be a strong one.

**Comment by tallclair** [comment]
> I recommend we propose this to beta but still off by default. What do folks think?

I'm against off-by-default betas on principle. What's the difference between an alpha and an off by default beta?

@sftim Can you raise your concerns on the exception request?

File: test/e2e/common/node/pod_resize.go
Comment (aojea) [inline test/e2e/common/node/pod_resize.go:952]: if is beta end enabled by default and does not depend on any external factor, you may remove `feature.InPlacePodVerticalScaling` so it gets exercised with all the common jobs ... because this feature will be exercised in all common clusters in prod

Comment (tallclair) [reply test/e2e/common/node/pod_resize.go]: Thanks. Done.

**Comment by tallclair** [comment]
/assign @thockin 
For API approval (fuzzer changes for roundtrip compatibility with defaulting)

**Comment by pacoxu** [comment]
Failure in `pull-kubernetes-e2e-kind` is related: Pod InPlace Resize Container `pod-resize-limit-ranger-test`. We may need to fix.

Re-run to confirm if it is a flake or failing test.

/test pull-kubernetes-e2e-kind

**Comment by tallclair** [comment]
Looks like a flake. I'll take a look at the test anyway.

**Comment by tallclair** [comment]
All blockers have been resolved! This is ready for review.
/hold cancel

Only other outstanding PR is https://github.com/kubernetes/kubernetes/pull/128754, which should not block the beta.

File: pkg/apis/core/fuzzer/fuzzer.go
Comment (thockin) [inline pkg/apis/core/fuzzer/fuzzer.go:313]: Given that this is a fuzzer, the chances of this happening is suuuuuuper slim.

Perhaps we should add something like:

* pick a random bool, if true set requests.cpu to a random value value between 100 and 10000 + "m"
* pick a random bool, if true set limits.cpu to prev.request + a random value between 0 and 1
* repeat for memory

That will give us a good probability of exercising the resources path.

Then here, set the policy to `[]string{ core.NotRequired, core.Required }[random int between 0 and 1]`
?

**Review by thockin — APPROVED** [review]
I'll approve for fuzzer, but I think it could be fuzzerier.  Minor

/approve

**Comment by aojea** [comment]
👏

**Comment by esotsal** [comment]
🙌

File: pkg/apis/core/fuzzer/fuzzer.go
Comment (liggitt) [inline pkg/apis/core/fuzzer/fuzzer.go:318]: if SetDefaults_Pod is defaulting new fields on container when this feature is enabled, won't that instant-restart pods on API server upgrade (https://github.com/kubernetes/kubernetes/issues/122028#issuecomment-1837528357) until the oldest node is 1.31 and includes the fix in https://github.com/kubernetes/kubernetes/pull/124220?

Comment (liggitt) [reply pkg/apis/core/fuzzer/fuzzer.go:318]: possible options to avoid the instant restart were to wait for 1.31 to be the oldest supported skewed node or to move the "defaulting" of the pod fields to PrepareForCreate (https://github.com/kubernetes/kubernetes/issues/122028#issuecomment-1838905110)

Comment (tallclair) [reply pkg/apis/core/fuzzer/fuzzer.go:318]: We should just remove the defaulting... Kubelet already handles the missing resize policy as an implicit default. I don't see any reason to set the explicit default.

Comment (liggitt) [reply pkg/apis/core/fuzzer/fuzzer.go:318]: I haven't followed all the discussions on the feature / API, but that seems likely to be the best option

**Pull request merged**
